app/src/main/java/org/simple/clinic/patient/PatientSearchResultDataSource.kt

```diff
@@ -55,11 +55,27 @@ class PatientSearchResultDataSource(
     val loadParamsForDatabaseSource = LoadRangeParams(params.startPosition, params.loadSize)
     source.loadRange(loadParamsForDatabaseSource, object : LoadRangeCallback<PatientSearchResult>() {
       override fun onResult(patientSearchResult: MutableList<PatientSearchResult>) {
+        val patientSearchResults = partitionSearchResultsByFacility(patientSearchResult, currentFacility.get())
+        val searchResultsItems = SearchResultsItemType.from(patientSearchResults)
```

Comment:
This will not work since `SearchResultsItemType#from` will only work correctly if we are loading the entire set of patient results at once. Here when it's paginated, since we are loading data in chunks, using this factory method will return the wrong set of results.

Consider the following scenario, where we have patients 1..10, the first five of which are in the current facility and the latter five are not.

Regardless of whether we are loading them in a paginated fashion or not, the complete set of results should look like this:

```
currentFacilityHeader
patient1
patient2
patient3
patient4
patient5
NotInCurrentFacilityHeader
patient6
patient7
patient8
patient9
patient10
```

Consider a scenario where we are loading search results in chunks of 3 patients at a time. What will happen with the paginated data source in the current implementation is:

```
currentFacilityHeader
patient1
patient2
patient3    ---> Page 1
currentFacilityHeader
patient4
patient5
NotInCurrentFacilityHeader
patient6    ---> Page 2
currentFacilityHeader
NoPatientsInCurrentFacility
NotInCurrentFacilityHeader
patient7
patient8
patient9    --> Page 3
currentFacilityHeader
NoPatientsInCurrentFacility
NotInCurrentFacilityHeader
patient10    --> Page 4
```

This will happen because the factory method in `SearchResultsItemType` has no concept of data being loaded in chunks and assumes that the set it receives is the complete set.

Reply:
Wait, I thought it would retain data as it loads 🤔

Reply:
Yep, as far as I can tell it would retain the loaded data and make it part of the paged list. If it's not present in the paged list it would try to load. So this might work?

Reply:
The problem is that the chunk of data that is loaded in the `loadRange` callback is being passed to the factory method, which assumes the list it receives the entire list of data so the additional headers that we compute and add to the subset again.

Reply:
> Yep, as far as I can tell it would retain the loaded data and make it part of the paged list. If it's not present in the paged list it would try to load. So this might work?

It does do this, but this is done after the `loadRange` callback will be invoked, which will then append to the current list.

Reply:
Ah okay. Without keeping a reference of the list inside the data source, this might be difficult to implement here then? Since we convert the data source to a paged list at the screen, is possible to do this there? or is there a better approach? 🤔

Reply:
I'm considering writing a `DataSource` base class to abstract away this kind of logic. Maybe we can think about this problem separately and come back tomorrow to figure out the best approach to take here.

Reply:
@vinaysshenoy @msasikanth let's get on a call to discuss the approach we are planning to follow here?